File: src/main/kotlin/codeacademy/bookingforum/app/comment/CommentService.java
package codeacademy.bookingforum.app.comment;

import codeacademy.bookingforum.app.configuration.ResponseObject;
import codeacademy.bookingforum.app.ecxeption.global.UnsatisfiedExpectationException;
import codeacademy.bookingforum.app.ecxeption.user.UserNotFoundException;
import codeacademy.bookingforum.app.post.Post;
import codeacademy.bookingforum.app.post.PostRepo;
import codeacademy.bookingforum.app.user.auth.UserAuth;
import codeacademy.bookingforum.app.user.auth.UserAuthRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;
import org.springframework.web.context.request.WebRequest;

import java.util.Collections;
import java.util.List;
import java.util.Objects;

@Service
public class CommentService {
    @Autowired
    CommentRepository commentRepo;
    @Autowired
    CommentMapper commentMapper;
    @Autowired
    UserAuthRepo userRepo;
    @Autowired
    PostRepo postRepo;

    public ResponseObject create(CommentDto comment, WebRequest request) {
        UserAuth user = userRepo.findById(comment.getUserId()).orElse(null);
        if (user == null) {
            throw new UserNotFoundException("User with id "+comment.getUserId()+" does not exist!");
        }
        checkAuth(user);
        commentRepo.save(commentMapper.fromDto(comment));

        return new ResponseObject(Collections.singletonList("Comment successfully published."), HttpStatus.CREATED, request);
    }

    public List<Long> getList(Long id) {
        Post post = postRepo.findById(id).orElse(null);
        if (post == null) {
            throw new UnsatisfiedExpectationException("Can't get comments of non-existing Post");
        }

        return commentRepo.findAllIdByPost(post.getId());
    }

    public CommentDto getById(Long id) {
        Comment comment = commentRepo.findById(id).orElse(null);
        if (comment == null) {
            throw new UnsatisfiedExpectationException("Comment with id "+id+" does not exist!");
        }

        return commentMapper.toDto(comment);
    }

    // Validate authorization, check whether user is requesting his own resource
    void checkAuth(UserAuth user) {
        UserDetails userDetails = (UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (!Objects.equals(userDetails.getUsername(), user.getUsername())) {
            throw new UnsatisfiedExpectationException("Requesting and requested user do not match!");
        }
    }
}

File: src/main/kotlin/codeacademy/bookingforum/app/comment/CommentService.kt
package codeacademy.bookingforum.app.comment

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class CommentService {
    @Autowired
    private lateinit var mapper: CommentMapper

    @Autowired
    private lateinit var repository: CommentRepository

    fun getAllComments(): List<CommentDto>? {
        return mapper.fromEntityListToDtoList( repository.findAll().toList() )
    }

    fun getCommentById(id: Long): CommentDto? {
        return mapper.fromEntityToDto( repository.findByIdOrNull(id))
    }

    fun postComment(commentDto: CommentDto?){
        mapper.fromDtoToEntity(commentDto).let { repository.save(it) }
    }

    fun updateComment(id: Long, commentDto: CommentDto?) {
        mapper.fromDtoToEntity(
            CommentDto(
                id = id,
                dateCommented = commentDto?.dateCommented,
                content = commentDto?.content,
                postId = commentDto?.postId,
                userId = commentDto?.userId,
                likes = commentDto?.likes,
                dislikes = commentDto?.dislikes,
            )
        ).let { repository.save(it) }
    }

    fun deleteComment(id: Long): String {
        return if(repository.existsById(id)){
            repository.deleteById(id)
            "Deleted user with id $id"
        }else{
            "Id not found"
        }
    }

}